services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,202 @@
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.northbound.messaging.kafka;
+
+import org.openkilda.messaging.Message;
+import org.openkilda.messaging.error.ErrorMessage;
+import org.openkilda.messaging.error.MessageException;
+import org.openkilda.messaging.info.ChunkedInfoMessage;
+import org.openkilda.messaging.info.InfoData;
+import org.openkilda.messaging.info.InfoMessage;
+import org.openkilda.northbound.messaging.MessageProducer;
+import org.openkilda.northbound.messaging.MessagingChannel;
+
+import com.google.common.collect.Lists;
+import org.apache.commons.collections4.map.PassiveExpiringMap;
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import javax.annotation.PostConstruct;
+
+/**
+ * Implementation of {@link MessagingChannel} for kafka.
+ */
+@Component
+public class KafkaMessagingChannel implements MessagingChannel {
+
+    private static final Logger logger = LoggerFactory.getLogger(KafkaMessagingChannel.class);
+
+    @Value("${northbound.messages.expiration.minutes}")
+    private int expiredTime;
+
+    /**
+     * Requests that are in progress of processing.
+     */
+    private Map<String, CompletableFuture<InfoData>> pendingRequests;
+    private Map<String, CompletableFuture<List<InfoData>>> pendingChunkedRequests;
+
+    /**
+     * Collects messages that are not related to any known chain.
+     */
+    private Map<String, ChunkedInfoMessage> unlinkedChunks;
+
+    /**
+     * Chains of chunked messages, it is filling by messages one by one as soon as the next linked message is received.
+     */
+    private Map<String, List<ChunkedInfoMessage>> messagesChains;
+
+    @Autowired
+    private MessageProducer messageProducer;
+
+    /**
+     * Creates storages, that are able to remove outdated messages and requests.
+     */
+    @PostConstruct
+    public void setUp() {
+        pendingRequests = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new ConcurrentHashMap<>());
```

Comment:
What will happen with future object stored on caller side? We are responsible to "close" it with meaningfuull exception thrown in case of timeout.

But this "silent" drop approach doesn't care about called waiting for a response on future object...

Reply:
Fixed